Document and modify how coq_makefile parses _CoqProject (#14558)
coq/coq: doc/sphinx/practical-tools/utilities.rst

line options <command-line-options>`). Collecting the list of files
and options is the job of the ``_CoqProject`` file.

A ``_CoqProject`` file may contain the following kinds of entries in any order,
separated by whitespace:
* Selected options of coqc, which are forwarded to it directly. Currently these
are ``-Q``, ``-I``, ``-R`` and ``-native-compiler``.
* ``-arg`` options for other options of coqc that don’t fall in the above set.
* Options specific to coq_makefile. Currently this is ``-docroot``.
* Paths to files belonging to the project.
* Comments, started by an unquoted ``#`` and going on until the end of the
line.

A simple example of a ``_CoqProject`` file follows:

::

    -R theories/ MyCode

>     -arg "-w all"

Reviewer: May I suggest to keep
```
-arg -w
-arg all
```
or transform it to
```
-arg -w -arg all
```
because this workaround is known to work well with build systems and IDEs.
See e.g. https://github.com/coq/coq/issues/5773 and related PRs & issues.
Perhaps this should be documented in the manual.

Please ignore my comment if this has been recently fixed in at least `coq_makefile`, CoqIDE, and ProofGeneral.

Author: In this example I tried to declare that "best practice" in a _CoqProject file might use double-quotes instead of unquoted arguments.

For coq_makefile and ProofGeneral, `-arg -w -arg all` and `-arg "-w all"` are equivalent (before and after this PR). The issue that CoqIDE had is probably fixed in this PR, but I haven’t tested it.

Reviewer 2: This should indeed be tested.

Author: Tested it and could verify that CoqIDE works. What I did boils down to:
```bash
mkdir -p /tmp/A/theories
cat > /tmp/A/theories/A.v << EOF
Test Default Goal Selector.
EOF
cat > /tmp/A/_CoqProject << EOF
-R theories/ T
-arg "-w default"
-arg "-set 'Default Goal Selector=!'"
theories/A.v
EOF
```
Then open `/tmp/A/theories/A.v` in CoqIDE and step through the file.

(``Qux_plugin``). This reduces the chances of begin unable to load two
distinct plugins because of a clash in their auxiliary module names.

.. _arbitrary_arguments:

> 

Author: Didn’t know that. Thanks for pointing it out.

Passing arbitrary arguments to coqc
+++++++++++++++++++++++++++++++++++
Using either ``-arg "-w all"`` or ``-arg -w -arg all`` its possible to pass the
option ``-w all`` to ``coqc``. For options like ``-set 'Default Goal
Selector=!'`` quotes are necessary: ``-arg "-set 'Default Goal Selector=!'"``.
It is impossible to pass a double quote using ``-arg``, only single quotes work

> in this situation.

Reviewer 3: Since the quoting rules are undocumented, one has to pass it like that, according to the doc.
This is why this PR is very welcome.

> in this situation.

Author: Currently its only possible to enclose stuff in double quotes, but not in apostrophes.

> in this situation.

Author: > I think this PR should also add a test in the test-suite (there is a dedicated directory for coq_makefile),

Testing for what?

> in this situation.

Author: > I don't see why a user would enter the awkward `-arg -w -arg all`. Unless there is a specific case where you need this.

I also think, that this is pretty weird syntax. But I think it’s actually better behaved passing `-arg "-w all"`, since otherwise generating `COQMF_OTHERFLAGS` becomes a lot more difficult. If we’d merge the argument parsing of coqc and coq_makefile we could remove the need for `-arg`, but I think such a thing should go into a separate PR.

.. _comments:

Comments
++++++++
The syntax for comments in a ``_CoqProject`` file is a bit unusual. The
character ``#`` starts a comment until the end of the line only if it is
outside a quoted string, and preceded by whitespace or a closing double quote.
This means ``Filename#.v`` and ``"Filename With # Spaces.v"`` won’t cause

> comments. The ``#`` is in these cases treated like an ordinary character.

Author: Since newline characters are whitespace, if `#` appears at the beginning of the line, it technically is "preceded by whitespace", thus starting a comment.
But I worded it a bit differently, so the text doesn’t have to explain this technicality.

> comments. The ``#`` is in these cases treated like an ordinary character.

Reviewer 4: (I think `#` as the first character of the file also starts a comment)

> comments. The ``#`` is in these cases treated like an ordinary character.

Author: Yes, I wasn’t too clear above.

.. _grammar_of_coqproject:

The grammar of _CoqProject
++++++++++++++++++++++++++
The parser of coq_makefile uses the following grammar:

  .. insertprodn CoqProject unquoted_string

> 

Author: Thanks for pointing this out. I was copying the style from some other file, without knowing what it meant.

  .. prodn::
     CoqProject ::= {* {| @blank | @comment | @quoted_string | @unquoted_string } }
     blank ::= {| space | horizontal_tab | newline }
     comment ::= # {* comment_byte } newline
     quoted_string ::= " {* quoted_string_byte } "
     unquoted_string ::= string_start_byte {* unquoted_string_byte }

Where the following definitions apply:

* :n:`space`, :n:`horizontal_tab` and :n:`newline` stand for the corresponding ASCII values.
* :n:`comment_byte` is the set of all bytes except :n:`newline`.
* :n:`quoted_string_byte` is the set of all bytes except ``"``.
* :n:`string_start_byte` is the set of all bytes except those that match :n:`@blank`, or are ``"`` or ``#``.
* :n:`unquoted_string_byte` is the set of all bytes except those that match :n:`@blank`.

The parser produces a list of strings in the same order as they were
encountered in ``_CoqProject``. All :n:`@blank` and :n:`@comment` tokens
get ignored and the double quotes of :n:`@quoted_string` tokens are removed as

> well.

Author: That was the reminder I kinda was expecting.
Thanks for the link.

> I expect `unquoted_string` is broken into multiple tokens.

I’m not sure what you mean, but I think not. An `unquoted_string` begins with a `string_start_byte` and goes on until the next `blank`. This whole resulting thing is one token. (That’s why `"` and `#` don’t have any special effects if they occur as `unquoted_string_byte`)

Maybe I can think of some wording that avoids introducing "byte", while still being technically correct. Maybe first pointing out, that _CoqProject files are always treated as if they were UTF-8 encoded, since then distinguishing "byte" from "character" is unnecessary, since all bytes with special semantics are UTF-8 characters and all other bytes UTF-8 encoded text get treated as "ordinary bytes".

The list of strings is then parsed using the following grammar:

  .. insertprodn options option3

  .. prodn::
     options ::= {* @option }
     option ::= {| @option0 | @option1 string | @option2 string string | @option3 string string string | string = string | -arg string | string }
     option0 ::= {| -impredicative-set | -no-install | -custom | -no-opt | -byte | -full | -opt }
     option1 ::= {| -install | -native-compiler | -f | -o }
     option2 ::= {| -Q | -I | -R }

>      option3 ::= {| -extra | -extra-phony }

Reviewer 5: Users want to think in terms of the available options.  Grouping by the number of values for each option is not how they will think about them.  Since you're not saying what these mean here, I don't think you need to list all the details.  This might be better just with prose rather than a grammar.

The grammar given above might accept a given list of strings in two different
ways, precisely if an equals-sign is involved. In this case the implementation

> chooses the leftmost interpretation.

Reviewer 5: Please be more precise (and simpler, if possible).  Sounds like the equals sign is the only special case.

Author: Jup, will do.

Author: The equals sign belongs to deprecated behaviour, and (probably) shouldn’t be used anymore. It can still be used to add variable definitions to CoqMakefile.local. The line `FOO = BAR` in _CoqProject produces `FOO = BAR` in CoqMakefile.local, where BAR may be an arbitrary token and FOO must not be a recognized option (except `-arg` which is allowed).

Author: I changed the code, so `-arg` and `_ = _` interact differently, giving `_ = _` less precedence.

:n:`string` stands for an arbitrary string and all other strings stand for

> themselves.

Reviewer 5: You should use something like `@string`, then you don't need this confusing sentence.  But use a different name, `string` is already defined elsewhere in the doc and may not be the same as what's accepted here.

Author: Thanks for explaining the syntax.
Hm, there is no place where the link could point to, because I nowhere define `string`. And instead of "string" we could use "token".

Reviewer 5: ```:token:`foo` ``` in regular text makes "foo" a nonterminal and shows it in boldface at the point of definition.  `foo` can then be referenced as `@foo` in prodn and `:n:` constructs.  (This is exactly how `@string` is defined--examine the other rst files if you wish.)

The semantics of ``-arg`` are as follows: all strings that were found
immediately to the right of ``-arg`` by the above grammars are concatenated
with single spaces between them. The resulting string is written in
``CoqMakefile.conf`` as variable ``COQMF_OTHERFLAGS``. This variable is then
passed to ``coqc`` using something similar to the following invocation: ``coqc
$(COQMF_OTHERFLAGS)``.

From this intermediate step in the ``CoqMakefile.conf``, a lot of the
peculiarities of ``-arg`` arise, since the string may be split appart
differently than how coq_makefile did it, variables like ``$(MAKE)`` can get
expanded and single quotes mark where no expansion or splitting should take

> place.

Reviewer 5: This is not too clear and not written from the user's perspective.

Reviewer 3: I think this PR should also change coq_makefile so that stuff like `-arg "some stuff"` is immediately desugared to `-arg some -arg stuff` so that this paragraph about nasty peculiarities can be omitted. The quotation rules will then be specific to coq_makefile, and will put make and dune out of the picture.

Author: > This is not too clear and not written from the user's perspective.

I agree and I’ll think about this, how it could be made easier to understand. Trouble is, that the current behaviour *just is* difficult to understand.

Author: @gares, how would you expect `-arg "some 'weird stuff' happens"` to behave? I think we clearly want to parse this as `["-arg", "some 'weird stuff' happens"]` (in coq_makefile, before splitting), but a naive implementation would pass the following list to coqc (after splitting): `["some", "'weird" "stuff'", "happens"]`, but I think `["some", "weird stuff", "happens"]` is more useful.

We could define, that single quotes only have effect inside the argument to `-arg`.

Reviewer 5: It's a judgement call whether and how to document bugs and limitations.  It's better to fix simple bugs rather than documenting them, but that's not always easy.  On the other hand, we do want to document limitations that may not be addressed soon (or ever).  Also, we don't want explanations of bugs and limitations to overshadow the basic description of the feature.

It would be great if you're willing to clean up the argument parsing a bit so it makes sense and is easier to describe and use.  For example, addressing the point you make [below ](https://github.com/coq/coq/pull/14558#issuecomment-869660724).

Author: So am I free to introduce, say, backslash escaping and propose a new parsing/output algorithm?

Reviewer 5: Feel free to propose any changes you wish, but they will have to be approved by reviewer(s).  Backslash escaping sounds like a small and compatible improvement--maybe just go ahead and code that.  Proposing a new parsing/output algorithm you might describe it first before writing code.

(``Qux_plugin``). This reduces the chances of begin unable to load two
distinct plugins because of a clash in their auxiliary module names.

Comments
++++++++
Comments begin with a ``#`` at the start of a line, after whitespace or after a

> closing double quote. They continue to the end of the current line. The

Reviewer 5: The part about "after a closing double quote" -- do you think that was a design intention or just a happenstance during coding?  If it's the latter we could consider whether to drop this feature in the interest of simplicity, which would have to be balanced against the expected impact on existing user code.

Author: I’d never have thought about this approach. Thanks for the input. I didn’t find any information about this, the code stems from https://github.com/coq/coq/commit/aad9936c225ef5b4f463b4af9ce1fa267bab337b.

The simplest behaviour for users would IMO be: "every `#` outside quotes starts a comment until the end of the current line". I don’t think a lot of users have strings with `#` in them, so I’ll do the cry-test and change the syntax accordingly.

We could mirror POSIX shell even more, by splitting `a "b"c` into `["a", "bc"]` instead of `["a", "b", "c"]` as coq_makefile does now. I think mirroring bash (or POSIX shell), is "very intuitive" behaviour for _CoqProject, since an unsuspecting user probably expects the same syntax in _CoqProject as in the shell.
This splitting is another reason why the passage you cited is so awkward.

Reviewer 5: Handling `"b"c` that way seems odd.  I'm not sure why Posix did it that way or if it's really good to duplicate that behavior.  (Just my initial thought--maybe there was a good reason for Posix to do that.  And others may like it.  Note I won't be the reviewer for these code changes.)

Author: Hm. I’m fine with the current quoting, as long as it is documented.

(``Qux_plugin``). This reduces the chances of begin unable to load two
distinct plugins because of a clash in their auxiliary module names.

Comments
++++++++
Every ``#`` outside of double quotes starts a comment until the end of the
line. Comments are ignored.

Quoting arguments to coqc
+++++++++++++++++++++++++
Any string in a ``_CoqProject`` file may be enclosed in double quotes to include
whitespace characters or ``#``. For example, use ``-arg "-w all"`` to pass the
string ``-w all`` as an argument to coqc. In an argument to `-arg`, single
quotes can appear within the double quotes if quoting twice is necessary. For
example ``-arg "-set 'Default Goal Selector=!'"`` gets passed to coqc as
``-set 'Default Goal Selector=!'``. But in all other circumstances, single

> quotes aren’t special characters.

Reviewer 5: Were you going to apply this or disagree with the suggestion?

Forbidden filenames
+++++++++++++++++++
The paths of files listed in a ``_CoqProject`` file may not contain any of the
following characters: ``\n``, ``\t``, space, ``\``, ``'``, ``"``, ``#``, ``$``,
``%``. These characters have special meaning in a Makefile and correctly

> encoding them isn't implemented in ``coq_makefile``.

Author: A technicality: Only paths of files are checked for special characters, but all other paths aren't. That’s why I chose the start of the paragraph like that.

For example `-R "foo $bar" Foo` is permitted by coqProject_file.ml and would not pose a problem for `make` if coq_makefile.ml would encode it correctly. *ahem* (coq_makefile.ml currently writes `COQMF_COQLIBS = -R 'foo $bar' Foo`, but in this case `make` expands the variable `$bar`. A correct solution would be `-R 'foo $$bar' Foo`.)

> encoding them isn't implemented in ``coq_makefile``.

Reviewer 5: OK.  If you like the wording I suggested for "These characters have ..." please include it.

Warning: No common logical root
+++++++++++++++++++++++++++++++
When a ``_CoqProject`` file contains something like ``-R theories Foo
theories/Bar.v``, the ``install-doc`` target will install the documentation
generated by ``coqdoc`` into ``user-contrib/Foo/``, in the folder where Coq was
installed.

But if the ``_CoqProject`` file contains something like:

::

    -R theories/Foo Foo
    -R theories/Bar Bar
    theories/Foo/Foo.v
    theories/Bar/Bar.v

the Coq files of the project don’t have a logical path in common and
``coq_makefile`` doesn’t know where to install the documentation. It will give
a warning: "No common logical root" and generate a Makefile that installs the
documentation in some folder beginning with "orphan", in the above example,

> it'd be ``user-contrib/orphan_Foo_Goo``.

Author: Oops, I changed the example from "Goo" to "Bar", but forgot to change this as well.

  One can append lines to the generated ``.merlin`` file extending this
  target.

.. note::

  For users of ``coq_makefile`` with version < 8.7

  + Support for "subdirectory" has been deprecated in version 8.7 and removed

>     in version 8.15. To perform actions before or after the build (like

Author: Yeah, I think removing this paragraph is ok. I moved it here from below, trying to put it in a place where it made more sense. But I think it isn’t needed.

  utility, in particular it may build a shared prerequisite twice.

The grammar of _CoqProject
++++++++++++++++++++++++++

> A _CoqProject file is implicitly assumed to be encoded in UTF-8. The parser of
> ``coq_makefile`` uses the following grammar:

Reviewer 5: ```suggestion
``coq_makefile`` uses the following grammar:
```

No need to mention UTF-8, either.

Author: The only difference in semantics between linebreaks and other blanks is, that linebreaks terminate comments. And this behaviour is, I think, already explained.

  .. prodn::
     CoqProject ::= {* {| @blank | @comment | @quoted_string | @unquoted_string } }
     blank ::= {| space | horizontal_tab | newline }

>      comment ::= # {* comment_byte } newline

Reviewer 5: ```suggestion
     comment ::= # {* comment_char } newline
```

Author: Thx

.. note::

  ``make foo.vo bar.vo -j`` has a different meaning for the ``make``
  utility, in particular it may build a shared prerequisite twice.

.. note::

  Due to limitations with the compilation chain, makefiles generated
  by ``coq_makefile`` won't correctly compile OCaml plugins with OCaml
  < 4.07.0 when using more than one job (``-j N`` for ``N > 1``).

The grammar of _CoqProject
++++++++++++++++++++++++++
The parser of ``coq_makefile`` uses the following syntax:

  .. prodn::
     CoqProject ::= {* {| @blank | @comment | @quoted_string | @unquoted_string } }
     blank ::= {| space | horizontal_tab | newline }
     comment ::= # {* comment_char } newline
     quoted_string ::= " {* quoted_char } "
     unquoted_string ::= string_start_char {* unquoted_char }

Where the following definitions apply:

* :n:`space`, :n:`horizontal_tab` and :n:`newline` stand for the corresponding
  ASCII characters.
* :n:`comment_char` is the set of all characters except :n:`newline`.
* :n:`quoted_char` is the set of all characters except ``"``.
* :n:`string_start_char` is the set of all characters except those that match :n:`@blank`, or are ``"`` or ``#``.

> * :n:`unquoted_char` is the set of all characters except those that match :n:`@blank` or are ``#``.

Reviewer 5: The grammar is not helpful.  For example, instead of

```
     comment ::= # {* comment_char } newline

* :n:`comment_char` is the set of all characters except :n:`newline`.
```

It's much more readable to say "Comments begin with a # and continue to the end of the line."

Please rework this part.

Author: Would it make more sense to move this grammar into coqProject_file.ml itself? Since I think such a grammar is a big help in understanding how [`parse`](https://github.com/Columbus240/coq/blob/bb7061659bc4dc074ba844bd231884586988112a/lib/coqProject_file.ml#L73-L124) works.

If the grammar were removed from the refman, I’d remove the rest of the paragraph as well, since that part can be inferred from the code.

Reviewer 5: That would be reasonable.

Author: Still thinking about this and how to do it.

Author: I couldn't format the BNF satisfactorily in ocamldoc and didn't come up with an alternative description of the grammar that I find adequate. Either the text came out imprecise, overly wordy or introducing more concepts.

I think it's fine leaving it as it is. The intuitive description of the syntax is spread over the other parts of the doc. So an ordinary user probably won't gain anything from the grammar-paragraph. But for devs of other programs that parse `_CoqProject`, a formal description is all kinds of useful. It would make the file format look a little less ad-hoc and will help finding and fixing differences in behaviour (for example in PG or find-bugs.py).

Reviewer 5: It could be ordinary comments rather than ocamldoc.  Unfortunately, it's very easy for formal descriptions to get out of date, so it's often necessary to verify the description is still current.  We can put this in for now, but it may be reworked when the section is eventually revised.

The parser produces a list of strings in the same order as they were

> encountered in ``_CoqProject``. All :n:`@blank` and :n:`@comment` tokens
> get ignored and the double quotes of :n:`@quoted_string` tokens are removed as
> well.

Reviewer 5: ```suggestion
encountered in ``_CoqProject``. Blanks and comments are removed
and the double quotes of :n:`@quoted_string` tokens are removed as
well.
```

IIUC

The current approach has a few limitations: Double quotes are only special
characters at the start of a string. And for lack of an escaping mechanism, it
is currently impossible to encode the following kinds of strings in a
``_CoqProject`` file:

* strings starting with ``"``
* strings starting with ``#`` and containing ``"``

> * strings containing both whitespace and ``"``

Author: I didn't want to decide how to parse the escape sequences, and when tokens should "stay together" or "split apart" on (non-)escaped whitespace. I think I'd first need a specification of how a new escaping/quoting mechanism should behave (as seen from the user's p.o.v.) before I can implement it.
Maybe discuss and resolve these limitations in a separate thread (issue/PR/cep/Zulip)? How about I sketch how the new parser might behave, then "we" (whoever is interested) can discuss that and in the end I implement it?

I think it would be best, if the new behaviour wouldn’t treat `-arg` as a special case, as it does now. (Regarding single-quotes)

About quoted/unquoted strings. I wanted to say something like: After splitting the _CoqProject file into tokens, it is impossible to have tokens of the following form: ...

> * strings containing both whitespace and ``"``

Reviewer 5: It's your choice whether to add an escape mechanism.  I'd consider how useful it is to users to support these before spending time on it--e.g. how often do users want a `"` at the beginning of a string (or escaped at all)?  You could create a short proposal as a new issue if you wish and perhaps include `@gares` in the description to get his attention.

Hope you will tweak this wording to be clearer about quoted/unquoted strings.

The list of strings is then processed as follows: If the first string of the
list is a known option, interpret the next strings as arguments of this option

> and remove these strings from the list. If the first string of the list is not

Reviewer 5: "next strings" - on the same line?

Unclear: "removing these strings" meaning it won't be passed to coqc?  The text should be clear about what is done with the resulting list of strings (i.e. it's passed to coqc IIUC).

Author: I tried to describe `process_cmd_line` using imperative wording. I'll try to think of better wording.

Reviewer 5: It's OK to use the passive voice at times.  The imperative form works when the subject is the reader, e.g. "[You] eat your broccoli"

Author: I wasn't clear. I meant "imperative vs. functional programming", not in the grammatical sense.
